sound/soc/sof/topology.c

```diff
@@ -2362,6 +2363,48 @@ int snd_sof_complete_pipeline(struct snd_sof_dev *sdev,
 	return 1;
 }
 
+static int snd_sof_get_kcontrol_val(struct snd_sof_dev *sdev)
```

Comment:
@lgirdwood a comment is added. The function is called in sof_complete. I believe that it will be only called in boot up. What I want to do is to set the initial value in the cache where store the values of kcontrols. So I think we need to read all the kcontrols.

Reply:
ok, the default kcontrol values should be set from topology loading or by userspace writing the value after soundcard int. i.e. alsactl

Reply:
Agree. I noticed the kcontrol value will be restore when we restore pipelines. So the kcontrol value will keep align when DSP is active. And the default value should be set by alsactl though I don't like to rely on userspace. So, I think we don't need to care about the default value on this PR, right?

Reply:
@lgirdwood So we don't need to deal with the default value on this PR, right?

Reply:
@bardliao the default value for each kcontrol will be cached when get() is called on it the first time after initial loading from topology. All ALSA kcontrols are only modified via RMW sequence giving us a chance to cache them all when they are first accessed by userspace.

Reply:
@lgirdwood Yes, and what the function do is very similar to the first read. Note that I removed the ipc functions from the ALSA kcontrol get() functions in this PR. So, it will get the cached data even when the get() is called first time. The snd_sof_get_kcontrol_val function will get data from DSP when firmware loading is finished and store it to the cache. And ALSA kcontrol get() function will get the right data from cache. But as you said, alsactl should set the default value for each kcontrol. So I think there will not be any issue if we remove this function.

Reply:
@bardliao we will need to keep an invalidate and writeback flag for each kcontrol which will indicate whether we need to invalidate and read back values from FW or writeback new value to firmware.

Reply:
@lgirdwood I update the PR according to your suggestion. Could you review it?